x-pack/plugins/monitoring/server/kibana_monitoring/bulk_uploader.js

```diff
@@ -0,0 +1,122 @@
+ * bulk endpoint.
+ *
+ * NOTE: internal collection will be removed in 7.0
+ *
+ * Depends on
+ *   - 'xpack.monitoring.kibana.collection.enabled' config
+ *   - monitoring feature from xpack_main.info isAvailable() and isEnabled()
+ * The dependencies are handled upstream
+ * - Ops Events - essentially Kibana's /api/status
+ * - Usage Stats - essentially Kibana's /api/stats
+ * - Kibana Settings - select uiSettings
+ * @param {Object} server HapiJS server instance
+ * @param {Object} xpackInfo server.plugins.xpack_main.info object
+ */
+export class BulkUploader {
+  constructor(server, xpackMainInfo, { interval, combineTypes }) {
+    if (typeof interval !== 'number') {
+      throw new Error('interval number of milliseconds is required');
+    }
+    if (typeof combineTypes !== 'function') {
+      throw new Error('combineTypes function is required');
+    }
+
+    this._timer =  null;
+    this._xpackMainInfo = xpackMainInfo;
+    this._interval = interval;
+    this._combineTypes = combineTypes;
+    this._log = getCollectorLogger(server);
+
+    this._client = server.plugins.elasticsearch.getCluster('admin').createClient({
+      plugins: [monitoringBulk],
+    });
+
+    this._callClusterWithInternalUser = callClusterFactory(server).getCallClusterInternal();
+
+  }
+
+  /*
+   * Start the interval timer
+   * @param {CollectorSet} collectorSet object to use for initial the fetch/upload and fetch/uploading on interval
+   * @return undefined
+   */
+  start(collectorSet) {
+    this._log.info('Starting monitoring stats collection');
+    if (this._timer) {
+      this._log.warn('BulkUploader timer already started');
+      return;
+    }
+
+    this._fetchAndUpload(collectorSet); // initial fetch
+    this._timer = setInterval(() => {
+      this._fetchAndUpload(collectorSet);
+    }, this._interval);
+  }
+
+  /*
+   * start() and stop() are lifecycle event handlers for
+   * xpackMainPlugin state changes
+   */
+  stop() {
+    this._log.info('Stopping monitoring stats collection');
+    clearInterval(this._timer);
+    this._timer = null;
+  }
+
+  /*
+   * @param {CollectorSet} collectorSet
+   * @return {Promise} - resolves to undefined
+   */
+  async _fetchAndUpload(collectorSet) {
+    // Before every fetch, check the license and make sure the bulk endpoint on the ES side is up
+    const mainMonitoring = this._xpackMainInfo.feature('monitoring');
+    const monitoringBulkEnabled = mainMonitoring && mainMonitoring.isAvailable() && mainMonitoring.isEnabled();
+
+    if (!monitoringBulkEnabled) {
+      // debug level to not spam the logs
+      this._log.debug(`Skipping fetch and upload of monitoring stats due to monitoring bulk endpoint not being available`);
+      return;
+    }
+
+    const data = await collectorSet.bulkFetch(this._callClusterWithInternalUser);
+    const usableData = data.filter(d => Boolean(d) && !isEmpty(d.result));
+    const payload = usableData.map(({ result, type }) => {
+      if (!isEmpty(result)) {
```

Comment:
I think it would also become more readable to just change this to

```js
const payload = data.filter(d => Boolean(d) && !isEmpty(d.result)).map(({ result, type }) => [{ index: { _type: type } }, result]);
```